action.yml

```diff
@@ -0,0 +1,117 @@
+        # ToDo: Re-enable cache when https://github.com/actions/setup-python/issues/361 is fixed
+        poetry_cache_enabled: 'false'
+
+    - name: Get Image URI
+      shell: bash
+      id: get-image-uri
+      run: |
+        # For testing purposes ( actual value: =master-${{ github.sha }} )
+        imageTag=$(echo ${{ github.ref_name }} | awk '{print tolower($0)}' | sed -e 's|/|-|g')
+        IMAGE_TAGS=$(
+          aws ecr describe-images \
+          --repository-name ${{ inputs.ecr_repository }} \
+          --image-ids imageTag="${imageTag}" \
+          --query 'imageDetails[*].imageTags[0]' \
+          --output json \
+          --region ${{ inputs.aws_region }}
+        )
+        REPOSITORY_URI=$(
+          aws ecr describe-repositories \
+          --repository-name ${{ inputs.ecr_repository }} \
+          --query 'repositories[0].repositoryUri' \
+          --output text \
+          --region ${{ inputs.aws_region }}
+        )
+        IMAGE_URI=$(jq --arg uri "${REPOSITORY_URI}" '.[] | ($uri + ":" + .)' <<< "${IMAGE_TAGS}")
+        echo "image-uri=${IMAGE_URI}" >> $GITHUB_OUTPUT
+
+    - name: Render and Deploy local-exec task definition to Amazon ECS
+      id: deploy-local-task-definition
+      uses: moneymeets/action-ecs-deploy/custom-deploy-steps@feature/MD-7037-implement-ecs-deployment-workflow
+      with:
+        application_id: ${{ inputs.ecr_repository }}-local-exec-${{ inputs.environment }}
+        image_uri: ${{ steps.get-image-uri.outputs.image-uri }}
+        aws_access_key_id: ${{ inputs.aws_access_key_id }}
+        aws_secret_access_key: ${{ inputs.aws_secret_access_key }}
+        aws_region: ${{ inputs.aws_region }}
+
+    - name: Render and Deploy task definition to Amazon ECS
+      id: deploy-task-definition
+      uses: moneymeets/action-ecs-deploy/custom-deploy-steps@feature/MD-7037-implement-ecs-deployment-workflow
+      with:
+        application_id: ${{ inputs.ecr_repository }}-${{ inputs.environment }}
+        image_uri: ${{ steps.get-image-uri.outputs.image-uri }}
+        aws_access_key_id: ${{ inputs.aws_access_key_id }}
+        aws_secret_access_key: ${{ inputs.aws_secret_access_key }}
+        aws_region: ${{ inputs.aws_region }}
+
+    - name: Update service
+      shell: bash
+      run: |
+        aws ecs update-service \
+          --health-check-grace-period-seconds 900 \
+          --task-definition "${{ steps.deploy-task-definition.outputs.latest-task-definition-arn }}" \
+          --cluster ${{ inputs.environment }} \
+          --service ${{ inputs.ecr_repository }}-${{ inputs.environment }} \
+          --desired-count 1 \
+          --region ${{ inputs.aws_region }}
+        
+        aws ecs wait services-stable \
+          --cluster ${{ inputs.environment }} \
+          --service ${{ inputs.ecr_repository }}-${{ inputs.environment }} \
+          --region ${{ inputs.aws_region }}
+
+    - name: Deregister previous local-exec task definition
+      if: always()
```

Comment:
I'm not sure that this is the correct logic to deal with deployment failures. If we assume the following state before deployment:

```
Active task definitions
 revision: 2  (tag = "GiHhub Actions Deployment") 
 revision: 1  (tag = "Pulumi") 
```

Then, during deployment we have

```
Active task definitions
 revision: 3  (tag = "GitHub Actions Deployment") 
 revision: 2  (tag = "GitHub Actions Deployment") 
 revision: 1  (tag = "Pulumi") 
```

The state after a successful deployment should be

```
Active task definitions
 revision: 3  (tag = "GitHub Actions Deployment") 
 revision: 1  (tag = "Pulumi") 
```

while for a failed deployment, we want

```
Active task definitions
 revision: 2  (tag = "GitHub Actions Deployment") 
 revision: 1  (tag = "Pulumi") 
```
So I think we need to correctly account for failed deployments, and not always remove the previous task definition.

Reply:
Discussed via call; True we should not always remove previous task definition but confirm (check) that the `service stable` step passes; only then deregister previous task task definition
If it fails; it then deregisters the latest task definition created in the earlier step

Reply:
The above logic also might not be enough for a success rollback, a `todo` added for the infrastructure part on pulumi is to look into the [Deployment circuit breaker](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/deployment-circuit-breaker.html) configuration for a service which in theory adds addition roll back mechanics in case of failed deployment

cc @felix11h @marns93

Reply:
Thank you for looking into this, @IamAbbey. Tomorrow, I will still create a ticket that we can link to the ToDo in the code, and once we have set up Circuit Breaker we can make sure task definitions are deregistered correctly here in case of rollbacks.

Reply:
I have created a ticket for this here, MD-7199.